keps/sig-api-machinery/20191001-oasv3.md

```diff
@@ -0,0 +1,949 @@
+            "type": "integer",
+            "format": "int64"
+          },
+          "resourceVersion": {
+            "description": "String that identifies the server's internal version of this object that can be used by clients to determine when objects have changed. Value must be treated as opaque by clients and passed unmodified back to the server. Populated by the system. Read-only. More info: https://git.k8s.io/community/contributors/devel/sig-architecture/api-conventions.md#concurrency-control-and-consistency",
+            "type": "string"
+          },
+          "selfLink": {
+            "description": "selfLink is a URL representing this object. Populated by the system. Read-only.\n\nDEPRECATED Kubernetes will stop propagating this field in 1.20 release and the field is planned to be removed in 1.21 release.",
+            "type": "string"
+          }
+        }
+      },
+      "meta.v1.Patch": {
+        "description": "Patch is provided to give a concrete name and type to the Kubernetes PATCH request body.",
+        "type": "object"
+      }
+    },
+    "parameters": {
+      "query.pretty.5ca316":  {
+        "name": "pretty",
+        "in": "query",
+        "description": "If 'true', then the output is pretty printed.",
+        "schema": {
+          "type": "string"
+        }
+      }
+    }
+  }
+}
+```
+
+### New names for objects
+
+The old spec uses the full path to a package as a name for an object, for example `ConfigMap` is defined as `io.k8s.api.core.v1.ConfigMap`.
+This not only doesn't add any value but also reveals the internal path to a type. 
+
+Instead the new names will contain information about a group, version and kind(`GVK`). `ConfigMap` in the new format would become `core.v1.ConfigMap`.
+
+### Reusable parameters
+
+In order to make the spec smaller and more clearer common parameters for operations will be kept under `components/parameters`. 
+The [OASv3 spec](https://github.com/OAI/OpenAPI-Specification/blob/master/versions/3.0.0.md#parameter-object) says that a unique parameter is defined by a combination of a name and location. 
+Thus to make a parameter unique across all operations its name will be derived as `name.location.SHA1(spec3.Parameter.MarshalJSON()) | hex.EncodeToString() | hexEncodedString[1:5]`.
+The last part takes the first 6 characters from a hex-encoded `SHA1` value that was calculated from a JSON encoded representation of a parameter e.g. `query.pretty.5ca316`.
```

Comment:
Won't we be at risk of collisions when using only the first 6 characters of a encoded `SHA1`?

Reply:
IIRC this is related to [the birthday problem](https://en.wikipedia.org/wiki/Birthday_problem). The chances of a collision can be calculated from `k^2/2N`. In this case, we could expect the first collision after generating `~5700` values. Which is enough for our needs.

Reply:
I was curious, so I checked and it looks like we only have ~35 distinct parameters. Probibly a couple more if we distinguish between dryRun from the different options types.

I'd be inclined to give the common ones clear names like `meta.v1.DeleteOptions/dryRun` to all the common ones and the just inline the less common ones and avoid the hashes altogether.

```
curl -s localhost:8085/openapi/v2 | jq '..|. | select(type=="object") | select(has("parameters")) | .parameters | select(type=="array")| .[].name' | sort | uniq -c | sort -rn
    452 "dryRun"
    452 "body"
    432 "pretty"
    365 "watch"
    365 "timeoutSeconds"
    365 "resourceVersion"
    365 "limit"
    365 "labelSelector"
    365 "fieldSelector"
    365 "continue"
    365 "allowWatchBookmarks"
    316 "fieldManager"
    233 "namespace"
    201 "name"
    136 "propagationPolicy"
    136 "orphanDependents"
    136 "gracePeriodSeconds"
    115 "force"
     61 "export"
     61 "exact"
      9 "path"
      3 "container"
      2 "tty"
      2 "stdout"
      2 "stdin"
      2 "stderr"
      1 "timestamps"
      1 "tailLines"
      1 "sinceSeconds"
      1 "previous"
      1 "ports"
      1 "logpath"
      1 "limitBytes"
      1 "follow"
      1 "command"
```